src/cocotb/share/lib/gpi/GpiCommon.cpp

```diff
@@ -161,6 +166,70 @@ static void gpi_load_libs(std::vector<std::string> to_load) {
     }
 }
 
+static int gpi_load_users() {
+    auto users = getenv("GPI_USERS");
+    if (!users) {
+        LOG_ERROR("No GPI_USERS specified, exiting...");
+    }
+    // I would have loved to use istringstream and getline, but it causes a
+    // compilation issue when compiling with newer GCCs against C++11.
+    std::string users_str = users;
+    std::string::size_type start_idx = 0;
+    bool done = false;
+    while (!done) {
+        auto next_delim = users_str.find(';', start_idx);
+        if (next_delim == std::string::npos) {
+            done = true;
+            next_delim = users_str.length();
+        }
+        auto user = users_str.substr(start_idx, next_delim - start_idx);
+        start_idx = next_delim + 1;
+
+        auto split_idx = user.rfind(',');
+
+        std::string lib_name;
+        std::string func_name;
+        if (split_idx == std::string::npos) {
+            lib_name = std::move(user);
+        } else {
+            lib_name = user.substr(0, split_idx);
+            func_name = user.substr(split_idx + 1, std::string::npos);
+        }
+
+        void *lib_handle = utils_dyn_open(lib_name.c_str());
+        if (!lib_handle) {
+            LOG_ERROR("Error loading library '%s'", lib_name.c_str());
+            gpi_finish();
```

Comment:
`gpi_finalize()` won't be called to run the finalize callbacks for the `GPI_USERS` that have already executed their entry functions.
That relies on `gpi_to_simulator()` which is only called when in a callback.

Reply:
Yeah this stuff is all very less than ideal. I'll add it to the backlog of stuff to fix. The more you try to fix the exponentially larger the diff becomes until you've rewritten everything. Then good luck getting it reviewed.